src/lib/FreeC/LiftedIR/Converter/Expr.hs

```diff
@@ -0,0 +1,253 @@
+--
+-- > ⎡Γ ⊢ p:Bool  Γ ⊢ t:τ  Γ ⊢ f:τ⎤'     Γ' ⊢ p':Bool'  Γ' ⊢ t':τ'  Γ' ⊢ f':τ'
+-- > ⎢----------------------------⎥ = -------------------------------------------
+-- > ⎣ Γ ⊢ if p then t else f : τ ⎦   Γ' ⊢ p' >>= λx:𝔹'.if x then t' else f' : τ'
+--
+-- Note that the argument of the lambda is lifted, but its type is
+-- @Bool Shape Pos@, which is just an alias for @bool@, which ignores its
+-- arguments.
+liftExpr' (IR.If srcSpan cond true false _) [] [] = do
+  cond' <- liftExpr cond
+  cond' `bind` \d -> LIR.If srcSpan d <$> liftExpr true <*> liftExpr false
+
+-- @case@-expressions.
+--
+-- > ⎡Γ ⊢ e:τ₀   Γ ⊢ alts:τ₀ => τ⎤'     Γ' ⊢ e':τ₀'     Γ' ⊢ alts':τ₀* => τ'
+-- > ⎢---------------------------⎥ = ------------------------------------------
+-- > ⎣  Γ ⊢ case e of alts : τ   ⎦   Γ' ⊢ e' >>= λx:τ₀*.match x with alts' : τ'
+--
+-- where @alts'@ are the lifted (not smart) constructors for τ₀.
+liftExpr' (IR.Case srcSpan discriminante patterns _) [] [] = do
+  discriminant' <- liftExpr discriminante
+  discriminant' `bind` \d -> LIR.Case srcSpan d <$> mapM convertAlt patterns
+
+liftExpr' (IR.Undefined srcSpan _    ) _ _ = return $ LIR.Undefined srcSpan
+
+liftExpr' (IR.ErrorExpr srcSpan msg _) _ _ = return $ LIR.ErrorExpr srcSpan msg
+
+-- Visible type application of an expression other than a function or
+-- constructor.
+liftExpr' expr (_ : _) _ =
+  reportFatal
+    $  Message (IR.exprSrcSpan expr) Internal
+    $  "Only type arguments of functions and constructors can be "
+    ++ "applied visibly."
+
+-- Application of an expression other than a function or constructor
+-- application. We use an as-pattern for @args@ such that we get a compile
+-- time warning when a node is added to the AST that we do not cover above.
+liftExpr' expr [] args@(_ : _) =
+  join $ generateApply <$> liftExpr expr <*> mapM liftExpr args
+
+-------------------------------------------------------------------------------
+-- Lift Patterns                                                             --
+-------------------------------------------------------------------------------
+
+-- | Translates a constructor pattern from IR to LIR.
+convertConPat :: IR.ConPat -> LIR.ConPat
+convertConPat (IR.ConPat srcSpan name) = LIR.ConPat srcSpan name
+
+-- | Translates a case alternative pattern from IR to LIR, by lifting the pattern
+--   and the right-hand side.
+convertAlt :: IR.Alt -> Converter LIR.Alt
+convertAlt (IR.Alt srcSpan conPat pats expr) = do
+  (pats', expr') <- convertAlt' pats expr
+  return $ LIR.Alt srcSpan (convertConPat conPat) pats' expr'
+
+-- | Translates an alternative (which consist of a list of variable patterns and
+--   a bound expression). Strict variables are replaced with fresh ones, which
+--   are unwrapped using @>>=@.
+convertAlt' :: [IR.VarPat] -> IR.Expr -> Converter ([LIR.VarPat], LIR.Expr)
+convertAlt' [] expr = ([], ) <$> liftExpr expr
+convertAlt' (pat@(IR.VarPat srcSpan name varType strict) : pats) expr = do
```

Comment:
These functions should be called `liftConPat`, `liftAlt` and `liftAlt'`, respectively.